### PR TITLE
[Pytorch-iOS] Enable QNNPACK for iOS

### PR DESCRIPTION
**Pull request by xta0**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24030 [Pytorch-iOS] Enable QNNPACK for iOS**

Summary:

The cmake arg - `USE_QNNPACK` was disabled for iOS build due to its lack of support for building multiple archs(armv7;armv7s;arm64) simultaneously.To enable it, we need to specify the value of IOS_ARCH explicitly in the build command:

```
./scripts/build_ios.sh \
-DIOS_ARCH=arm64 \
-DBUILD_CAFFE2_MOBILE=OFF \
```
However,the iOS.cmake will overwirte this value according to the value of `IOS_PLATFORM`. This PR is a fix to this problem.

Test Plan:

- `USE_QNNPACK` should be turned on by cmake.
- `libqnnpack.a` can be generated successfully.
- `libortch.a` can be compiled and run successfully on iOS devices.

Differential Revision: [D16771014](https://our.internmc.facebook.com/intern/diff/D16771014)

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
@xuhdev if you wanna take a look :)

**Review by xuhdev — COMMENTED** [review]
Is `IOS_ARCH` supposed to be an optional accessible to the user? If so, why not make it an option and make its default value depending on something determined by `ISO_PLATFORM`? If not, I don't see why we need `if (NOT DEFINED ISO_ARCH)`.

**Review by ezyang — CHANGES_REQUESTED** [review]
see @xuhdev's comments

**Comment by xta0** [comment]
> Is `IOS_ARCH` supposed to be an optional accessible to the user? If so, why not make it an option and make its default value depending on something determined by `ISO_PLATFORM`? If not, I don't see why we need `if (NOT DEFINED ISO_ARCH)`.

It is an optional accessible to the user. You can specify which ARCH you want by using a sub-shell arg or a cmake arg

```
./scripts/build_ios.sh \
-DIOS_ARCH=arm64 \
-DBUILD_CAFFE2_MOBILE=OFF \

IOS_ARCH=arm64 ./scripts/build_ios.sh \
-DBUILD_CAFFE2_MOBILE=OFF \
```
However, the `iOS.cmake` will be loaded first, the `IOS_ARCH` will be overwritten by the default value of `IOS_PLATFORM`, thus in the following cmake commands, the value `IOS_ARCH` is changed. For example, the `Dependencies.cmake` will check `IOS_ARCH` to decide if QNNPACK should be disabled or not. You can try run the commands above and log `IOS_ARCH` to see the value.

**Comment by xuhdev** [comment]
I understand that you can pass in the build option, but `IOS_ARCH` is not cached in CMakeCache.txt. That means subsequent builds may have that information lost, and CMake GUI tools will not recognize it as a build option. To make it a proper build option, you need to set a [cache variable](https://cmake.org/cmake/help/latest/command/set.html#set-cache-entry). You can refer to `CMakeLists.txt` for some examples.

**Comment by xta0** [comment]
> I understand that you can pass in the build option, but `IOS_ARCH` is not cached in CMakeCache.txt. That means subsequent builds may have that information lost, and CMake GUI tools will not recognize it as a build option. To make it a proper build option, you need to set a [cache variable](https://cmake.org/cmake/help/latest/command/set.html#set-cache-entry). You can refer to `CMakeLists.txt` for some examples.

You mean cache this `IOS_ARCH` in the root `CMakeLists.txt`? I think all iOS specific logic should be kept in `iOS.cmake`. Could you be more specific? Maybe share some code you think is the best way to do it?

**Comment by xuhdev** [comment]
It doesn't have to be in `CMakeLists.txt` -- I pointed you to it because it has shown some examples. You can add something like the following to iOS.cmake:

```cmake
set(IOS_ARCH "${DEFAULT_COMPUTED_FROM_ISO_PATFORM}" CACHE STRING "Arch of the iOS build.")
```

**Comment by xta0** [comment]
> It doesn't have to be in `CMakeLists.txt` -- I pointed you to it because it has shown some examples. You can add something like the following to iOS.cmake:
> 
> ```cmake
> set(IOS_ARCH "${DEFAULT_COMPUTED_FROM_ISO_PATFORM}" CACHE STRING "Arch of the iOS build.")
> ```

Thanks for the example, I can add that cache code for sure. But without the `if-else` check in this PR, the value still gonna be overridden, isn't it?

**Comment by ezyang** [comment]
You don't need the if-else. See the documentation for `set( ... CACHE)`:

> Sets the given cache <variable> (cache entry). Since cache entries are meant to provide user-settable values this does not overwrite existing cache entries by default. Use the FORCE option to overwrite existing entries.

File: cmake/iOS.cmake [hunk 1]
```diff
@@ -159,13 +159,14 @@ set (CMAKE_OSX_SYSROOT ${CMAKE_IOS_SDK_ROOT} CACHE PATH "Sysroot used for iOS su
...
 # set the architecture for iOS 
 if (IOS_PLATFORM STREQUAL "OS")
-    set (IOS_ARCH "armv7;armv7s;arm64")
+    set (DEFAULT_IOS_ARCH "armv7;armv7s;arm64")
 elseif (IOS_PLATFORM STREQUAL "SIMULATOR")
-    set (IOS_ARCH "i386;x86_64")
+    set (DEFAULT_IOS_ARCH "i386;x86_64")
 elseif (IOS_PLATFORM STREQUAL "WATCHOS")
-    set (IOS_ARCH "armv7k")
+    set (DEFAULT_IOS_ARCH "armv7k")
```
Comment (xuhdev) [inline cmake/iOS.cmake]: Would you also like to make `IOS_PLATFORM` a build option? Did you also miss an `else` here for fatal errors?

Comment (xta0) [reply cmake/iOS.cmake]: You don't have to specify `IOS_PLATFORM` explicitly, the default one is `OS`.

**Comment by xta0** [comment]
> You don't need the if-else. See the documentation for `set( ... CACHE)`:
> 
> > Sets the given cache  (cache entry). Since cache entries are meant to provide user-settable values this does not overwrite existing cache entries by default. Use the FORCE option to overwrite existing entries.

Talked to Edward, this is he suggested 

```
if (IOS_PLATFORM STREQUAL "OS")
    set (DEFAULT_IOS_ARCH "armv7;armv7s;arm64")
elseif (IOS_PLATFORM STREQUAL "SIMULATOR")
    set (DEFAULT_IOS_ARCH "i386;x86_64")
elseif (IOS_PLATFORM STREQUAL "WATCHOS")
    set (DEFAULT_IOS_ARCH "armv7k")
endif ()

set (IOS_ARCH ${DEFAULT_IOS_ARCH} CACHE string  "Build architecture for iOS")
```

**Review by xuhdev — APPROVED** [review]

File: cmake/iOS.cmake (hunk 1 above)
Comment (xuhdev) [reply cmake/iOS.cmake]: What I meant is that `IOS_PLATFORM` may be set with a typo (like `QATCHOS`). What would happen in this case?

(I realized that you did not introduce these in this PR, so I think it's fine to leave it in this PR.)

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**